apps/cowswap-frontend/src/modules/tradeWidgetAddons/containers/MetamaskTransactionWarning/index.tsx

```diff
@@ -40,3 +62,31 @@ export function MetamaskTransactionWarning({ sellToken }: { sellToken: Currency
     </Banner>
   )
 }
+
+function useWidgetProviderMetaInfo() {
+  const provider = useWalletProvider()
+  const [widgetProviderMetaInfo, setWidgetProviderMetaInfo] = useState<ProviderMetaInfoPayload | null>(null)
+
+  const rawProvider = provider?.provider as unknown
+
+  useEffect(() => {
+    const isWidgetEthereumProvider = rawProvider instanceof WidgetEthereumProvider
+
+    if (!isWidgetEthereumProvider) return
+
+    rawProvider.onProviderMetaInfo(setWidgetProviderMetaInfo)
+  }, [rawProvider])
+
+  return widgetProviderMetaInfo
+}
+
+/**
+ * This is hacky way to detect if the wallet is metamask mobile injected wallet
+ * Many injected wallet browsers emulate isMetaMask, but only metamask mobile has _metamask
+ */
+function useIsMetamaskMobileInjectedWallet(): boolean {
+  const walletProvider = useWalletProvider()
+  const rawProvider = walletProvider?.provider as any
```

Comment:
I think we should not make use of `as`, if something we could use typescript type guards